### ISSUE TITLE
Feature Request : Add a resetable filament counter into EEPROM

### ISSUE THREAD

**Issue by nicolas-rambaud** [MarlinFirmware/Marlin#924]
Hello, 
I think it could be a cool feature to add a counter of "used filament length" accessible from LCD and where the data would be accumulated into EEPROM.
Some LCD/GCode command/menu could reset the filament counter at user will.

Like this we can have some indication on how many meter we have left on our spool like the DaVinci 3D printer.
Sometimes it's not very clear if we have enough filament left to print a big or a small  plastic part.

To me the easiest way to do it is to just count the length of filament used and sum it up at the end of the print or after a print abort . The user will know either calculation or after the use of 1 full plastic spool, how many meters there is on a 3mm/2.85mm/1.75mm spool.
Based on that he will know approximately how many meters he has left on his own spool.

Please tell me what you think about it.

**Comment by Ziggy2013** [comment]
The running total of the filament used during a print is already available in Marlin via the global variable

current_position[E_AXIS]

However the slicers (eg slic3r and Cura) insert a G92 E0 gcode at intervals with the aim of preventing overflow (or rounding errors) in the current_position[E_AXIS] value.  Slic3r inserts a G92 very regularly whereas Cura only inserts a G92 if the total filament count is over 9999mm.

It would easily be possible to maintain an accurate running total of filament consumed by using current_position[E_AXIS] and adding that value to an overall total whenever a G92 is received. I have modded this in my own firmware version and it works well.  The total filament used can also be reset to zero (or stored) whenever a G28 is received (on the assumption a new print is starting)

```
case 92: // G92
  if (!code_seen(axis_codes[E_AXIS])) st_synchronize();
  for (int8_t i=0; i < NUM_AXIS; i++) {
    if (code_seen(axis_codes[i])) {
      if (i == E_AXIS) {
        Total_E[E_AXIS] = Total_E[E_AXIS] + current_position[E_AXIS]; // update running total
        current_position[i] = code_value();
        Total_E[E_AXIS] = Total_E[E_AXIS] - code_value(); // if value is <> 0
        plan_set_e_position(current_position[E_AXIS]);
      }
    }
```

The total  filament used value is also very useful when determining how long it will take to complete a print.  It turns out that the filament is used at a reasonably constant rate throughout the print.  By extrapolating the filament used over time versus total filament required it is possible to get a reasonably accurate estimate of the time required to complete the print.

**Comment by thinkyhead** [comment]
I like it! And this value could be shown along with total print time when the print completes. The current completion line is over-long anyway.

**Comment by Ziggy2013** [comment]
@thinkyhead To be able to calculate a time to complete using the filament consumed, Marlin also needs to know the total filament required.  I think an M gcode will be required to do this early in the print.  The total filament could be calculated by scanning the print gcode file on the SD Card - but this is a fairly complex task.

**Comment by thinkyhead** [comment]
@Ziggy2013 Pre-scanning the GCode has been suggested before, but of course you can only do that with SD printing. Some slicers will include the total filament volume in a comment in the GCode. Within the firmware the only reliable way to get the measurement is to add it all up during printing and show the final value at the end. Print time estimation can be done throughout printing by looking at the total time elapsed and dividing it into the percentage completed.

**Comment by Ziggy2013** [comment]
@thinkyhead Absolutely agree - only reliable way to track filament consumed is to add it up during printing - most of the functionality to do this is already there.

Print time estimation using the "percentage complete" (ie the % of the SD Card file read versus the length of the file) is wildly inaccurate and effectively useless in practice.

However the filament consumed versus total filament required is far more accurate as a predictor of print finish time.

**Comment by thinkyhead** [comment]
I agree the time estimate is almost pointless. But I personally try to do this in my head all the time, and I wouldn't mind if the firmware did it for me.

**Comment by Ziggy2013** [comment]
Using filament consumed versus filament required over time gives an estimate of time to finish within around 5 mins over a 1-2 hour print and can compensate/adjust the time estimate if the feedrate is changed up/down during the print.

On my set up I have added an M gcode M601 Fxxxxx which tells the firmware the total filament xxxx required. The time estimate to finish then then becomes quite accurate and useful. And this works for both PC and SD based printing. Maybe I should raise a separate feature request for this.

Anyway having a running total of the total filament consumed will be a very good start.

**Comment by thinkyhead** [comment]
@Ziggy2013 I do like the idea of having a GCode that can convey this kind of information up-front. Of course, it would be up to host software to use it, or slicers at least. And, I presume many hosts like to do this themselves while monitoring the print.

**Comment by Ziggy2013** [comment]
@thinkyhead The M code to set the total filament required is very easy to implement.  However I am not aware of a standard M code that should be used. I decided to use M601 as there was a relationship with M600 "filament change"

M601 Ex Fyyyy ;  E = extruder #, yyyy filament required.

Another option would be to add the extruder# and filament required parameters to the existing M code

M701 ; Load Filament

IMO the printer should be told how much filament for each extruder will be required to complete the print. Probably should be told what type/color of filament to use as well.  All info future printers are likely to make use of.

**Comment by thinkyhead** [comment]
Well, to be fair, there's a lot of liberty taken in Marlin concerning the use of gcode. Is `M851` a standard? Nope, but it's used as a default code for Set Z Probe Offset. So we can be a little creative. Maybe there's a standard gcode for sending "info" to the robot. I'll poke around the GCode documentation on RepRap wiki and see if there is such a thing…

**Comment by alexborro** [comment]
Nice feature.

But I still vote to scanning the GCode to sum up used filament.

Most host software actually show remaining time..

Cheers.

Alex
Em 27/03/2015 20:48, "Scott Lahteine" notifications@github.com escreveu:

**Comment by thinkyhead** [comment]
@alexborro SD Card printing only, of course. But I'd be curious to see how long it takes for my Arduino to read through an entire GCode file from the SD card. I know this interface is verrrrry slowwww.

**Comment by scotty1024** [comment]
I really can't see reading a 113MB g-code file being practical much less parsing it reliably, totaling positive and negative filament moves, on a 16MHz processor reading the SD card at 1 million bits per second.. My napkin says just reading the file much less doing the scanning would take 20 minutes.

**Comment by thinkyhead** [comment]
I'd be curious to see what Cura, Slic3r, and others are sticking into the top of the GCode they create now. In past versions they would include comments about the total filament volume. It would be possible to parse the comments and take note of these values, when available, although this leans towards falling outside the point of Marlin, which is meant to be a more-or-less "dumb" controller, leaving duties like these up to host software, and reading GCode _agnostically_.

I realize there are a few good hosts already out there, like Pronterface, Repetier Host, and the Cura front-end. But would it make any sense to build a Marlin Host – a host specifically tailored to accompany Marlin, which could act as a reference implementation for other hosts?

**Comment by alexborro** [comment]
@thinkyhead,  would be it possible to read the file during print? I don't
know whether we can have to pointer to a single file or not... If yes, we
could do it in parallel and present the estimated time as soon as we scan
all the file.

Cheers.

Alex.
Em 27/03/2015 22:20, "Scott Lahteine" notifications@github.com escreveu:

**Comment by scotty1024** [comment]
Here is Simplify3D

http://pastebin.com/raw.php?i=k1d15NKf http://pastebin.com/raw.php?i=k1d15NKf

Here is Cura:

M190 S100.000000
M109 S235.000000
;Sliced at: Sun 22-02-2015 07:41:18
;Basic settings: Layer height: 0.25 Walls: 0.8 Fill: 35
;Print time: 30 minutes
;Filament used: 2.562m 7.0g
;Filament cost: None
;M190 S100 ;Uncomment to add your own bed temperature line
;M109 S235 ;Uncomment to add your own temperature line
G21 ; units to millimeters
M107 ; Turn off fan
G28 X0 Y0 ; Home extruder
G28 Z0 ; auto-level?
G90 ; Absolute positioning
G92 E0
M82 ; Extruder in absolute mode
;Put printing message on LCD screen
M117 Printing...

;Layer count: 130
;LAYER:0

**Comment by thinkyhead** [comment]
@alexborro Theoretically it should be possible. But the main drawback to reading additional data during the print is that Marlin _probably_ isn't very well set up to deal with two file handles at once, and if we use only a single file handle, then every read operation would need to seek, read, and then seek again to go back to the main input stream. I have a feeling all this extra work would wreak havoc on the smoothness of printing. It's certainly worth some experimentation to see how much it degrades performance on modest hardware.

**Comment by thinkyhead** [comment]
@scotty1024 Ah, so Cura is calculating that for us now. Very nice. Of course it calculates based on running the printer at the full speed specified in the GCode, and if you change the speed up or down it will need to recalculate the remaining time, most likely losing accuracy. It also can't possible account for the small but cumulative variation that comes from Acceleration and Firmware Retraction settings in the firmware…. It ends up becoming a pretty complicated affair to generate a meaningful value.

**Comment by alexborro** [comment]
Ok, so we need to rely on Slicer information.. I guess most of them publish
this info in gcode, we just need to parse different strings for each
Slicer.
Em 28/03/2015 01:11, "Scott Lahteine" notifications@github.com escreveu:

**Comment by scotty1024** [comment]
@thinkyhead But Cura also supplies filament length to use the suggested tracking of how much has been used during the print to have a highly accurate presentation of how far the printer is into the print.

**Comment by thinkyhead** [comment]
@scotty1024 Yes, that is a decent metric, but of course it still doesn't correlate very well with the time required to extrude that filament. Infill is usually faster than perimeters, etc., so regions with more infill will take less time.

**Comment by Ziggy2013** [comment]
Yes. Using the filament consumed versus total filament required has its limits as a predictor of estimated time to finish (ETF) a print. But in practice the ETF is reasonably accurate and far better than using file bytes read versus total size of the gcode file.

Anyway this issue is primarily about keeping track of the filament used - and that IMO is a very worthwhile improvement which can be done very accurately in Marlin as the gcode is read and actioned.

**Comment by boelle** [comment]
reading through this one the initial poster have not given any word for a year....

And i question how many would benefit from this held against the the work it requires... a host like pronterface tells how much filament is needed to do the print and it have never failed on me... i could add a sticky label on each reel and note how much have been used... but i just check manual if the required lenght is on the spool before print... if not i will postpone the print and order a new spool or change colour if it does not matter

will close this one as i belive it will not benefit the greater number of users
